Required Tests

- make-snv-peptides

  - General
    - Are resulting mutant peptides accurate?
    - Are resulting wildtype (patient, not reference) peptides accurate?
    - Is the information in the FASTA headers (both MT and WT) accurate?
    - WT (patient, not reference) sequence will need to be able to support heterozygotes.
  - Unit tests
    - Are the emitted lengths accurate?
    - Does the genomic context of the MT nt sequence matched the aa sequence?
  - Improvements
    - Currently it emits a length such that the shortest valid length (8mer)
      includes the changed amino acid of interest.
    - Use germline VCF in order to create germline peptide for agretopicity
      calculation.
    - The HGVS variant integration code should be an independent of functions.
      Should be able to accept a snpEff record(s) and apply to transcript.
    - Expand to include genomic context as part of header. This is only
      applicable to MT peptide.
    - Header should include a key:value approach to storing metadata so that
      the resulting information can be easily retrieved by the metadata function.

- get-snv-genomic-context
  // This could probably be incorporated into the header of the MT FASTA files
  // generated by make-snv-peptides

  - General
    - Are the nt sequences being fetched accurate?
    - Are the nt sequences being correctly applied to the correct header?
  - Unit tests
    - Are the emitted nt sequences the required length?
    - Do the emitted nt sequences align correctly to the aa coordinates?
  - Improvements
    - Perhaps it'd be better to take flanking sequence _around_ the peptide
      sequence rather than have the user define the overall length.
    - This needs to accept and utilize the annoated germline VCF. Currently,
      the "variant-free zone" around somatic SNVs make it less of a concern,
      but ideally want to be able to consider all missense SNVs.

- expressed-variants
  // This should probably be renamed to something like
  // "filter-expressed-variants". Basically, the naming conventions within
  // LensTools should follow a logical scheme, e.g. "filter-_",
  // "make-_-peptides", "add-\*-metadata".

  - General
    - Are the thresholds (e.g. percentile _and_ threshold) working as expected?
    - Are other metrics working as expected? (e.g., other than TPM)
    - Does the exclusion of zeros work as expected?
  - Unit tests
    - Given a known quant file and a set of percentiles, are the current thresholds reached?
    - Given a known quant file, does the threshold filter work correctly?
    - Test the above two tests but after excluding zeros.
    - Test the above three tests but with a non-TPM metric from the quant file.
  - Improvements
    - Update name.
    - Could be useful to know how many transcripts a variant is expression in
      and also the other tissues that express that variant. The number of
      expressed variants could likely be easily calculated in the
      add-snv-metadata mode.

- rna-covered-variants
  // This needs a better name. Perhaps "check-rna-coverage".

  - General
    - Have to consider how we want this to contribute towards other outputs.
      Currently it accepts a somatic VCF, a tumor BAM, and a coverage
      requirement. This can simply emit a VCF (in the same way expressed-variants and
      isolated-variants does). Maybe useful to also emit a metadata file for adding
      to the eventual metadata report.
  - Unit tests
  - Improvements
    - Add coverage for SNVs, disruptive indels and frameshift variants.
    - If a variant is provided, filter the variant. If a FASTA is provided, get
      the nt sequence (and surrounding context) of peptide of interest and
      ensure it's _observed_ in the RNA bam. THe FASTA mode should also append the
      observed depth of the peptide of interest.

- isolated-variants
  // This needs a better name. Perhaps "filter-isolated-variants".

  - General
    - Needs to be expanded with annotated germline VCF to allow silent hom/het
      or missense hom proximal germline mutations.
  - Unit tests
  - Improvements
    - See general. basically need to expand filtering to allow germline
      variants (even though I think it's a relative small number of variants
      affected).

- calculate-agretopicity
  // This seems pretty straight forward. May be nice to ramp up the internal
  // sanity checking to ensure the correct MT and WT peptides are being compared.

  - General
  - Unit tests
    - Is the agretopicity metric (mt binding affinity/wt binding affinity) okay?
  - Improvements
    - May not be a bad idea to plot a histgram of the agretopicity values for QC purposes.
    - May be useful to rename to calculate-snv-agretopocity so it's clear.
    - May be nice to have a warning/error in cases of agretopicity of 1 since
      it's likely a situation where the MT peptide does not actually contain
      any different amino acids.

- make-pyclone-vi-inputs
  // This currently utilizies a MuTect-2 specific VCF files in order to calculate
  // the ref and alt depths. Ideally these values should be averaged with the
  // Strelka2 and Abra2 Cadabra VCFs though maybe those aren't necessary as I
  // believe MuTect2 is the strictest of the three (given the AF VCF, 1000g VCF,
  // etc.).

  - General
  - Unit tests
  - Improvements
    - May be able to use depths from other variant callers. I imagine the
      depths should all be effectively the same though (assuming they aren't
      doing any internal normalization as they're all looking at the same input
      BAMs). The major benefit of using the other variant callers is that we
      are currently missing variants that are not present within MuTect2's VCFs.

- add-snv-metadata
  // This is probably already pretty comprehensive. May want to change it
  // retreiving data from the genomic context file to getting these data from the
  // mutant peptide FASTA.

  - General
  - Unit tests
  - Improvements
    - Include information regarding evidence of mutant peptide in
      RNA-Sequencing data. We can't confirm the reading frame from the
      RNA-Sequencing data, but still make sure the nt sequence exists.

- add-indel-metadata
  // This is already pretty comprehensive. Need to add genomic context.

  - General
  - Unit tests
  - Improvements
    - Include information regarding evidence of mutant peptide in
      RNA-Sequencing data. We can't confirm the reading frame from the
      RNA-Sequencing data, but still make sure the nt sequence exists.

- expressed-hervs
  // This should be likely renamed to filter-expressed-hervs so the naming
  // convention is more consistent.

  - General
  - Unit tests
  - Improvements
    - Ensure filtering percentiles are working appropriately.

- make-herv-peptides

  - General
    - This is using code that incorporates the bcftools workiflow (bcftools
      mpileup/bcftools call/bcftools consensus) in order to create a
      "patient-specific" sequence. This should probably include filtering for
      any germline variants called. Also, this workflow already accomodates the
      inclusion of heterozygous sites within the peptide creation phase.
  - Unit tests
  - Improvements
    - Eventually use an homology-based approach instead of all ORFs.

- add-herv-metadata

  - General
    - This should include everything it needs currently. May be nice to pull
      some information from HERVd if appropriate.
  - Unit tests
  - Improvements

- consolidate-multiqc-stats
  // This mode isn't super relevant to the current push for LENS on IO datasets.
  // Postponing any improvements for now.

  - General
  - Unit tests
  - Improvements

- expressed-self-genes
  // This should behave as expected, more or less. The counts include all
  // transcripts in the GTF, so should have no issue checking for expression of
  // transcripts from these genes of interest. May be worth switching name to have a
  // more consistent naming scheme, e.g. expressed-self-genes =>
  // filter-expresse-self-genes.

  - General
    - Are the thresholds (e.g. percentile _and_ threshold) working as expected?
    - Are other metrics working as expected? (e.g., other than TPM)
    - Does the exclusion of zeros work as expected?
    - Does the output ONLY include transcripts belonging to the genes of interest?
  - Unit tests
    - Given a known quant file and a set of percentiles, are the current thresholds reached?
    - Given a known quant file, does the threshold filter work correctly?
    - Test the above two tests but after excluding zeros.
    - Test the above three tests but with a non-TPM metric from the quant file.
  - Improvements

- make-self-antigen-peptides
  // This should work as expected, assuming peptide creation is working. The
  // current approach may be a bit wonky (full peptide sequences for any transcripts
  // with no mutations, sliding window for transcripts _with_ mutations. It's worth
  // including information regarding the germline mutations that may be observed on
  // a peptide-level. This information could be directly included in the
  // self-antigen metadata report so users have more information regarding the
  // sequenced used for peptide generation. Should we be including somatic variants
  // here? I can see reason to and not to, but maybe it's best they're included
  // along with proper CCF data.
  - General
  - Unit tests
  - Improvements
    - Include information regarding evidence of _germline_ peptide in
      RNA-Sequencing data. We can't confirm the reading frame from the
      RNA-Sequencing data, but still make sure the nt sequence exists.
- add-self-antigen-metadata
  // Expand to include information regarding the sequence used to generate the
  // peptide. Useful information can be the number and type of germline and somatic
  // variants affecting each peptide.

  - General
  - Unit tests
  - Improvements

- filter-virdetect-by-counts
  // This should probablyt be renamed to filter-expressed-viruses. The current
  // naming convention assumes virdetect will be used, but this isn't a guanrantee.

  - General
  - Unit tests
  - Improvements

- make-viral-peptides

  - General
  - Unit tests
  - Improvements

- add-viral-metadata

  - General
  - Unit tests
  - Improvements

- add-fusion-peptides
  // This is effectively adding the content from STARFusion. Overall, this should
  // be find as-is.

  - General
  - Unit tests
  - Improvements

- add-fusion-metadata
  // It may be handy to include information regarding the sort of the fusion
  // (e.g. What algorithm predicted the fusion event?). It already has a lot of
  // relatively useful information.
  - General
    Does the metadata assigned to each peptide correspond to the same inputs from STARFusion predicted fusions file?
  - Unit tests
  - Improvements

TO ADD

- make-splice-peptides
  // This is already handled by NeoSplice, but would be nice to have
  // functionality that combined all of the outputs together and presents them in
  // a format similar to the other reports.

- add-splice-metadata
  // Probably already pretty comprehensive coming from NeoSplice, so will
  // start off as a pass-through function, but will be good to have it in
  // place for when we want to expand the outputs for NeoSplice.
